crypto/modes/asm/aes-gcm-avx512.pl

```diff
@@ -0,0 +1,5141 @@
+  my $LAST_GHASH_BLK  = $_[12];    # [in] XMM with last cipher text block shuffled for GHASH
+  my $ZT0             = $_[13];    # [clobbered] ZMM temporary
+  my $ZT1             = $_[14];    # [clobbered] ZMM temporary
+  my $ZT2             = $_[15];    # [clobbered] ZMM temporary
+  my $ZT3             = $_[16];    # [clobbered] ZMM temporary
+  my $ZT4             = $_[17];    # [clobbered] ZMM temporary
+  my $ZT5             = $_[18];    # [clobbered] ZMM temporary
+  my $ZT6             = $_[19];    # [clobbered] ZMM temporary
+  my $ZT7             = $_[20];    # [clobbered] ZMM temporary
+  my $ZT8             = $_[21];    # [clobbered] ZMM temporary
+  my $PBLOCK_LEN      = $_[22];    # [in] partial block length
+  my $GH              = $_[23];    # [in] ZMM with hi product part
+  my $GM              = $_[24];    # [in] ZMM with mid prodcut part
+  my $GL              = $_[25];    # [in] ZMM with lo product part
+
+  my $rndsuffix = &random_string();
+
+  if ($INSTANCE_TYPE eq "single_call") {
+
+    # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+    # ;;; SINGLE CALL case
+    # ;;; - hash all data including partial block
+    # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+    if (scalar(@_) == 23) {
+
+      # ;; start GHASH compute
+      &GHASH_1_TO_16($GCM128_CTX, $HASH_IN_OUT, $ZT0, $ZT1, $ZT2, $ZT3,
+        $ZT4, $ZT5, $ZT6, $ZT7, $ZT8, &ZWORD($HASH_IN_OUT), $DAT0, $DAT1, $DAT2, $DAT3, $NUM_BLOCKS);
+    } elsif (scalar(@_) == 26) {
+
+      # ;; continue GHASH compute
+      &GHASH_1_TO_16($GCM128_CTX, $HASH_IN_OUT, $ZT0, $ZT1, $ZT2, $ZT3, $ZT4, $ZT5,
+        $ZT6, $ZT7, $ZT8, &ZWORD($HASH_IN_OUT), $DAT0, $DAT1, $DAT2, $DAT3, $NUM_BLOCKS, $GH, $GM, $GL);
+    }
+
+  } else {
+
+    # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+    # ;;; MULTI CALL (SGL) case
+    # ;;; - hash all but the last partial block of data
+    # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+    # ;; update data offset
+    if ($NUM_BLOCKS > 1) {
+
+      # ;; The final block of data may be <16B
+      $code .= "sub               \$16 * ($NUM_BLOCKS - 1),$LENGTH\n";
+    }
+
+    if ($NUM_BLOCKS < 16) {
+      $code .= <<___;
+        # ;; NOTE: the 'jl' is always taken for num_initial_blocks = 16.
+        # ;;      This is run in the context of GCM_ENC_DEC_SMALL for length < 256.
+        cmp               \$16,$LENGTH
+        jl            .L_small_initial_partial_block_${rndsuffix}
+
+        # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+        # ;;; Handle a full length final block - encrypt and hash all blocks
+        # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+        sub               \$16,$LENGTH
+        movq              \$0,($PBLOCK_LEN)
+___
+
+      # ;; Hash all of the data
+      if (scalar(@_) == 23) {
```

Comment:
If INSTANCE_TYPE is removed, this 23 should be changed to 22 (same with 26 to 25 below)